datafusion/core/src/physical_plan/topk/mod.rs

```diff
@@ -0,0 +1,516 @@
+    fn k_largest(&self) -> Option<&TopKRow> {
+        if self.inner.len() < self.k {
+            None
+        } else {
+            self.inner.last()
+        }
+    }
+
+    /// Adds `row` to this heap. If inserting this new item would
+    /// increase the size past `k`, removes the previously smallest
+    /// item.
+    fn add(&mut self, batch_entry: &mut RecordBatchEntry, row: OwnedRow, index: usize) {
+        assert!(self.inner.len() <= self.k);
+
+        batch_entry.uses += 1;
+
+        self.owned_row_bytes += owned_row_size(&row);
+
+        // put the new row into the correct location to maintain that
+        // self.inner is sorted in descending order
+        let insertion_point = self
+            .inner
+            .partition_point(|current_row| current_row.row <= row);
+        self.inner.insert(
+            insertion_point,
+            TopKRow {
+                row,
+                batch_id: batch_entry.id,
+                index,
+            },
+        );
+
+        // limit size to k items
+        if self.inner.len() > self.k {
+            // If there was a previous minimum value, decrement its use
+            if let Some(prev_min) = self.inner.pop() {
+                if prev_min.batch_id == batch_entry.id {
+                    batch_entry.uses -= 1;
+                } else {
+                    self.store.unuse(prev_min.batch_id);
+                }
+                // update memory accounting
+                let prev_size = owned_row_size(&prev_min.row);
+                assert!(self.owned_row_bytes >= prev_size);
+                self.owned_row_bytes -= prev_size;
+            }
+        }
+    }
+
+    /// Returns the values stored in this heap, from values low to high, as a single
+    /// [`RecordBatch`]
+    pub fn emit(&self, schema: SchemaRef) -> Result<RecordBatch> {
+        // Indicies for each row within its respective RecordBatch
+        let indicies: Vec<_> = self
+            .inner
+            .iter()
+            .enumerate()
+            .map(|(i, k)| (i, k.index))
+            .collect();
+
+        let num_columns = {
+            let Some(first_value) = self.inner.get(0) else {
+                return Ok(RecordBatch::new_empty(schema));
+            };
+            self.store
+                .get(first_value.batch_id)
+                .expect("invalid batch id")
+                .batch
+                .num_columns()
+        };
+
+        // build the output columns one at time, using the
+        // `interleave` kernel to pick rows from different arrays
+        let output_columns: Vec<_> = (0..num_columns)
+            .map(|col| {
+                let input_arrays: Vec<_> = self
+                    .inner
+                    .iter()
+                    .map(|k| {
+                        let entry =
+                            self.store.get(k.batch_id).expect("invalid stored batch id");
+                        entry.batch.column(col) as &dyn Array
```

Comment:
Seems like until the very end we keep around the full batches that contain some of the top-k rows. This means that there is a potential edge case whereby the top 1000 rows are all distributed across 1000 different batches, thus negating the memory benefits to this operator.

Can we extract/keep track of only the `Row`s that are in the top-k (maybe that way the `BinaryHeap` could be used as well)? Or alternatively, perform Array slicing and splicing eagerly, as we iterate over the batches?

Reply:
This is an excellent point @gruuya 

>  Or alternatively, perform Array slicing and splicing eagerly, as we iterate over the batches?

I think this is a great idea, and I think it would be relatively straight forward to "compact" the RecordBatches that are held in the `RecordBatchStore` -- namely copy all the stored rows into a new, single RecordBatch.

I'll give this a try as well over the next day or two